src/main/scala/mesosphere/marathon/core/election/CuratorElectionStream.scala

```diff
@@ -64,23 +61,75 @@ object CuratorElectionStream extends StrictLogging {
       }
   }
 
-  private class CuratorEventEmitter(
-      metrics: Metrics,
-      singleThreadEC: ExecutionContext,
-      clientCloseable: LifeCycledCloseableLike[CuratorFramework],
-      zooKeeperLeaderPath: String,
-      hostPort: String,
-      sq: SourceQueueWithComplete[LeadershipState]) extends Cancellable {
+  private def leaderLatchStream(
+    metrics: Metrics,
+    clientCloseable: LifeCycledCloseableLike[CuratorFramework],
+    zooKeeperLeaderPath: String,
+    hostPort: String,
+    singleThreadEC: ExecutionContext): Source[LeadershipState, Cancellable] = {
+
+    Source.queue[LeadershipState](16, OverflowStrategy.dropHead)
+      .mapMaterializedValue { sq =>
+        val client = clientCloseable.closeable
+        val cancelled = new AtomicBoolean(false)
+        val curatorLeaderLatchPath = zooKeeperLeaderPath + "-curator"
+        val latch = new LeaderLatch(client, curatorLeaderLatchPath, hostPort)
 
-    val client = clientCloseable.closeable
-    private lazy val oldLeaderHostPortMetric: Timer =
+        /**
+          * On first invocation, synchronously close the latch.
+          *
+          * This allows us to remove our leadership offering before closing the zookeeper client.
+          */
+        lazy val closeHook: () => Unit = { () =>
+          if (cancelled.compareAndSet(false, true)) {
+            clientCloseable.removeBeforeClose(closeHook)
+            try {
+              logger.info("Closing leader latch")
+              latch.close()
+              logger.info("Leader latch closed")
+            } catch {
+              case ex: Throwable =>
+                logger.error("Error closing CuratorElectionStream latch", ex)
+            }
+            sq.complete()
+          }
+        }
+
+        try {
+          /* We register the beforeClose hook to ensure that we have an opportunity to remove the latch entry before we
+           * lose our connection to Zookeeper */
+          clientCloseable.beforeClose(closeHook)
+
+          logger.info("starting leader latch")
+          latch.start()
+          startLeaderEventsPoll(latch, metrics, singleThreadEC, client, curatorLeaderLatchPath, hostPort, sq)
+        } catch {
+          case ex: Throwable =>
+            logger.error("Error starting curator leader latch")
+            sq.fail(ex)
+        }
+
+        sq.watchCompletion().onComplete { _ => closeHook() }(singleThreadEC)
+        new CancellableOnce(closeHook)
+      }
+  }
+
+  private def startLeaderEventsPoll(
+    latch: LeaderLatch,
+    metrics: Metrics,
+    singleThreadEC: ExecutionContext,
+    client: CuratorFramework,
+    curatorLeaderLatchPath: String,
+    hostPort: String,
+    sq: SourceQueueWithComplete[LeadershipState]): Unit = {
+
+    require(latch.getState == LeaderLatch.State.STARTED)
```

Comment:
what do you think about writing a custom message within this require? I think it really improves the quality of logs and debuggability